File: app/src/main/java/dev/ragnarok/fenrir/adapter/LogsAdapter.java
package dev.ragnarok.fenrir.adapter;

import static dev.ragnarok.fenrir.util.Utils.safeLenghtOf;

import android.view.View;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import java.util.EventListener;
import java.util.List;

import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.adapter.base.RecyclerBindableAdapter;
import dev.ragnarok.fenrir.model.LogEvent;
import dev.ragnarok.fenrir.model.LogEventWrapper;
import dev.ragnarok.fenrir.util.AppTextUtils;

public class LogsAdapter extends RecyclerBindableAdapter<LogEventWrapper, LogsAdapter.Holder> {

    private static final int MAX_BODY_LENGHT = 400;
    private final ActionListener actionListener;

    public LogsAdapter(List<LogEventWrapper> data, ActionListener actionListener) {
        super(data);
        this.actionListener = actionListener;
    }

    @Override
    protected void onBindItemViewHolder(Holder holder, int position, int type) {
        LogEventWrapper wrapper = getItem(position);
        LogEvent event = wrapper.getEvent();

        holder.body.setText(event.getBody());
        holder.tag.setText(event.getTag());

        long unixtime = event.getDate() / 1000;
        holder.datetime.setText(AppTextUtils.getDateFromUnixTime(unixtime));

        holder.buttonShare.setOnClickListener(v -> actionListener.onShareClick(wrapper));
        holder.buttonCopy.setOnClickListener(v -> actionListener.onCopyClick(wrapper));

        holder.bodyRoot.setOnClickListener(v -> {
            if (!canReduce(event.getBody())) {
                return;
            }

            wrapper.setExpanded(!wrapper.isExpanded());
            notifyItemChanged(position + getHeadersCount());
            //setupBodyRoot(holder, wrapper);
        });

        setupBodyRoot(holder, wrapper);
    }

    private boolean canReduce(String body) {
        return safeLenghtOf(body) > MAX_BODY_LENGHT;
    }

    private void setupBodyRoot(Holder holder, LogEventWrapper wrapper) {
        String body = wrapper.getEvent().getBody();

        boolean canReduce = canReduce(body);

        if (!canReduce || wrapper.isExpanded()) {
            holder.buttonExpand.setVisibility(View.GONE);
            holder.body.setText(body);
        } else {
            holder.buttonExpand.setVisibility(View.VISIBLE);
            holder.body.setText(AppTextUtils.reduceText(body, MAX_BODY_LENGHT));
        }
    }

    @Override
    protected Holder viewHolder(View view, int type) {
        return new Holder(view);
    }

    @Override
    protected int layoutId(int type) {
        return R.layout.item_log;
    }

    public interface ActionListener extends EventListener {
        void onShareClick(LogEventWrapper wrapper);

        void onCopyClick(LogEventWrapper wrapper);
    }

    static class Holder extends RecyclerView.ViewHolder {

        final TextView tag;
        final TextView datetime;
        final TextView body;

        final View buttonShare;
        final View buttonCopy;

        final View bodyRoot;
        final View buttonExpand;

        Holder(View itemView) {
            super(itemView);
            tag = itemView.findViewById(R.id.log_tag);
            datetime = itemView.findViewById(R.id.log_datetime);
            body = itemView.findViewById(R.id.log_body);

            buttonShare = itemView.findViewById(R.id.log_button_share);
            buttonCopy = itemView.findViewById(R.id.log_button_copy);

            bodyRoot = itemView.findViewById(R.id.log_body_root);
            buttonExpand = itemView.findViewById(R.id.log_button_expand);
        }
    }

}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/fragment/logs/LogsAdapter.kt
package dev.ragnarok.fenrir.fragment.logs

import android.view.View
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.fragment.base.RecyclerBindableAdapter
import dev.ragnarok.fenrir.model.LogEventWrapper
import dev.ragnarok.fenrir.orZero
import dev.ragnarok.fenrir.util.AppTextUtils
import dev.ragnarok.fenrir.util.Utils

class LogsAdapter(data: MutableList<LogEventWrapper>, private val actionListener: ActionListener) :
    RecyclerBindableAdapter<LogEventWrapper, LogsAdapter.Holder>(data) {
    override fun onBindItemViewHolder(viewHolder: Holder, position: Int, type: Int) {
        val wrapper = getItem(position)
        val event = wrapper.getEvent()
        viewHolder.body.text = event?.body
        viewHolder.tag.text = event?.tag
        val unixtime = event?.date.orZero() / 1000
        viewHolder.datetime.text = AppTextUtils.getDateFromUnixTime(unixtime)
        viewHolder.buttonShare.setOnClickListener { actionListener.onShareClick(wrapper) }
        viewHolder.buttonCopy.setOnClickListener { actionListener.onCopyClick(wrapper) }
        viewHolder.bodyRoot.setOnClickListener {
            if (!canReduce(event?.body)) {
                return@setOnClickListener
            }
            wrapper.setExpanded(!wrapper.isExpanded())
            notifyItemChanged(position + headersCount)
        }
        setupBodyRoot(viewHolder, wrapper)
    }

    private fun canReduce(body: String?): Boolean {
        return Utils.safeLenghtOf(body) > MAX_BODY_LENGTH
    }

    private fun setupBodyRoot(holder: Holder, wrapper: LogEventWrapper) {
        val body = wrapper.getEvent()?.body
        val canReduce = canReduce(body)
        if (!canReduce || wrapper.isExpanded()) {
            holder.buttonExpand.visibility = View.GONE
            holder.body.text = body
        } else {
            holder.buttonExpand.visibility = View.VISIBLE
            holder.body.text = AppTextUtils.reduceText(body, MAX_BODY_LENGTH)
        }
    }

    override fun viewHolder(view: View, type: Int): Holder {
        return Holder(view)
    }

    override fun layoutId(type: Int): Int {
        return R.layout.item_log
    }

    interface ActionListener {
        fun onShareClick(wrapper: LogEventWrapper)
        fun onCopyClick(wrapper: LogEventWrapper)
    }

    class Holder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val tag: TextView = itemView.findViewById(R.id.log_tag)
        val datetime: TextView = itemView.findViewById(R.id.log_datetime)
        val body: TextView = itemView.findViewById(R.id.log_body)
        val buttonShare: View = itemView.findViewById(R.id.log_button_share)
        val buttonCopy: View = itemView.findViewById(R.id.log_button_copy)
        val bodyRoot: View = itemView.findViewById(R.id.log_body_root)
        val buttonExpand: View = itemView.findViewById(R.id.log_button_expand)
    }

    companion object {
        private const val MAX_BODY_LENGTH = 400
    }
}